# Getting Data ready for analysis

---


# Pre-Processing 
---

1. Get the Dataset
2. Importing Libraries
3. Importing the Dataset
4. Dealing with data types
  
5. Deal with Missing Data or other issues
6. Deal with Categorical Data
7. Save the file

# 1 Get the Dataset

About the data:

A driver uses an app to track GPS coordinates as he drives to work and back each day. The app collects the location and elevation data. Data for about 200 trips are summarized in this data set.

The original data can be downloaded from: [OpenMV.net](https://openmv.net/info/travel-times) Datasets.

This data wil be used for pre-processing purposes only. 




# Importing Libraries

In [1]:
import pandas as pd
import numpy as np

# Importing the Dataset

In [2]:
# Data Preprocessing

#A driver uses an app to track GPS coordinates as he drives to work and back each day. The app collects the location and elevation data. Data for about 200 trips are summarized in this data set.

# Importing the dataset
dataset = pd.read_csv("./travel-times.csv")
print("Data Shape: (Rows, Columns)", dataset.shape)

dataset.head()

Data Shape: (Rows, Columns) (205, 13)


Date StartTime  DayOfWeek GoingTo  Distance  MaxSpeed  AvgSpeed  \
0  1/6/2012     16:37     Friday    Home     51.29     127.4      78.3   
1  1/6/2012     08:20     Friday     GSK     51.63     130.3      81.8   
2  1/4/2012     16:17  Wednesday    Home     51.27     127.4      82.0   
3  1/4/2012     07:53  Wednesday     GSK     49.17     132.3      74.2   
4  1/3/2012     18:57    Tuesday    Home     51.15     136.2      83.4   

   AvgMovingSpeed FuelEconomy  TotalTime  MovingTime Take407All Comments  
0            84.8         NaN       39.3        36.3         No      NaN  
1            88.9         NaN       37.9        34.9         No      NaN  
2            85.8         NaN       37.5        35.9         No      NaN  
3            82.9         NaN       39.8        35.6         No      NaN  
4            88.1         NaN       36.8        34.8         No      NaN

In [3]:
print(dataset.columns)
dataset.dtypes

Index(['Date', 'StartTime', 'DayOfWeek', 'GoingTo', 'Distance', 'MaxSpeed',
       'AvgSpeed', 'AvgMovingSpeed', 'FuelEconomy', 'TotalTime', 'MovingTime',
       'Take407All', 'Comments'],
      dtype='object')


Date               object
StartTime          object
DayOfWeek          object
GoingTo            object
Distance          float64
MaxSpeed          float64
AvgSpeed          float64
AvgMovingSpeed    float64
FuelEconomy        object
TotalTime         float64
MovingTime        float64
Take407All         object
Comments           object
dtype: object

The dataset appears to have wrong datatypes.  We need to change it to its proper datatype.  For instance, some objects need to be converted to numeric types and "Date" needs to be converted from object type to numberic and then to datatime object. 
Let's continue below to do the conversios.

# Data Preprocessing

## Change data Proper Data Type

In [4]:
# changing the objects to numerics.s
dataset = dataset.convert_objects(convert_numeric=True)

/Users/Create/anaconda3/envs/machine-learn/lib/python3.6/site-packages/ipykernel_launcher.py:2: FutureWarning: convert_objects is deprecated.  Use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  


In [5]:
#checking datatypes 
dataset.dtypes

Date               object
StartTime          object
DayOfWeek          object
GoingTo            object
Distance          float64
MaxSpeed          float64
AvgSpeed          float64
AvgMovingSpeed    float64
FuelEconomy       float64
TotalTime         float64
MovingTime        float64
Take407All         object
Comments           object
dtype: object

In [6]:
# converting the Date column from an object to a datetime type.

dataset['Date'] = pd.to_datetime(dataset['Date'])
dataset.dtypes

Date              datetime64[ns]
StartTime                 object
DayOfWeek                 object
GoingTo                   object
Distance                 float64
MaxSpeed                 float64
AvgSpeed                 float64
AvgMovingSpeed           float64
FuelEconomy              float64
TotalTime                float64
MovingTime               float64
Take407All                object
Comments                  object
dtype: object

In [7]:
dataset = dataset.replace("-", "NaN")
dataset.head()

Date StartTime  DayOfWeek GoingTo  Distance  MaxSpeed  AvgSpeed  \
0 2012-01-06     16:37     Friday    Home     51.29     127.4      78.3   
1 2012-01-06     08:20     Friday     GSK     51.63     130.3      81.8   
2 2012-01-04     16:17  Wednesday    Home     51.27     127.4      82.0   
3 2012-01-04     07:53  Wednesday     GSK     49.17     132.3      74.2   
4 2012-01-03     18:57    Tuesday    Home     51.15     136.2      83.4   

   AvgMovingSpeed  FuelEconomy  TotalTime  MovingTime Take407All Comments  
0            84.8          NaN       39.3        36.3         No      NaN  
1            88.9          NaN       37.9        34.9         No      NaN  
2            85.8          NaN       37.5        35.9         No      NaN  
3            82.9          NaN       39.8        35.6         No      NaN  
4            88.1          NaN       36.8        34.8         No      NaN

# Dealing with Categorical Values

---

Before we get this data ready for modeling it is necessary to convert some categorical values to a numeric form. In a way, we will be converting categorical variables to indicator variables or dummy variables.  A dummy variable takes on the values 1 or 0.  

Here we will be using the map function because it simple and we have more control in specifying wich number belongs to what categorical value.


We wil use the .unique() function which will return unique values in the object. Uniques are returned in order of appearance, this does NOT sort.  This is due to see how many values correspond to each columns.

In [8]:
#let's check the uniqueness of DayOfWeek columnm
dataset['DayOfWeek'].unique()

array(['Friday', 'Wednesday', 'Tuesday', 'Monday', 'Thursday'], dtype=object)

We only have five days, So, we will need to include the five days and the five respecive numbers associated with each day. 

In [9]:
# Usign the .map function to assign a numerical value to each categorical value in the 
# DayOfWeek column

dataset['DayOfWeek'] = dataset['DayOfWeek'].map({'Friday':4, 'Wednesday': 2,
                                                 'Tuesday':1, 'Monday':0, 'Thursday':3})

Great, we have done changing to numerical values for the five day of the week.  Now, let's do the same for the others.

In [10]:
#let's check the unique values of GoingTo columnm

dataset['GoingTo'].unique()

array(['Home', 'GSK'], dtype=object)

We only have two values here, so let's change them in the same way we did above.

In [11]:
#Using the .map function to assign a numerical value to each categorical value in the 
#GoingTo column

dataset['GoingTo'] = dataset['GoingTo'].map({'Home':1, 'GSK': 0})

In [12]:
# let's check the unique values of 'Take407All' column

dataset['Take407All'].unique()

array(['No', 'Yes'], dtype=object)

In [13]:
#Using the .map function to assign a numerical value to each categorical value in the 
#'Take407All' column

dataset['Take407All'] = dataset['Take407All'].map({'No':1, 'Yes': 0})

Great,  we have changed all the categorical values to numerical values.

In [14]:
# Let's check our values 
dataset.head()

Date StartTime  DayOfWeek  GoingTo  Distance  MaxSpeed  AvgSpeed  \
0 2012-01-06     16:37          4        1     51.29     127.4      78.3   
1 2012-01-06     08:20          4        0     51.63     130.3      81.8   
2 2012-01-04     16:17          2        1     51.27     127.4      82.0   
3 2012-01-04     07:53          2        0     49.17     132.3      74.2   
4 2012-01-03     18:57          1        1     51.15     136.2      83.4   

   AvgMovingSpeed  FuelEconomy  TotalTime  MovingTime  Take407All Comments  
0            84.8          NaN       39.3        36.3           1      NaN  
1            88.9          NaN       37.9        34.9           1      NaN  
2            85.8          NaN       37.5        35.9           1      NaN  
3            82.9          NaN       39.8        35.6           1      NaN  
4            88.1          NaN       36.8        34.8           1      NaN

# Dealing with Nan values

In [15]:
dataset.isnull().sum()

Date                0
StartTime           0
DayOfWeek           0
GoingTo             0
Distance            0
MaxSpeed            0
AvgSpeed            0
AvgMovingSpeed      0
FuelEconomy        19
TotalTime           0
MovingTime          0
Take407All          0
Comments          181
dtype: int64

We have 19 NaN values in the FuelEconomy column and 181 NaN values in the Comments column.

In [16]:
dataset['Comments'].fillna(0, inplace=True)
dataset.head()

Date StartTime  DayOfWeek  GoingTo  Distance  MaxSpeed  AvgSpeed  \
0 2012-01-06     16:37          4        1     51.29     127.4      78.3   
1 2012-01-06     08:20          4        0     51.63     130.3      81.8   
2 2012-01-04     16:17          2        1     51.27     127.4      82.0   
3 2012-01-04     07:53          2        0     49.17     132.3      74.2   
4 2012-01-03     18:57          1        1     51.15     136.2      83.4   

   AvgMovingSpeed  FuelEconomy  TotalTime  MovingTime  Take407All Comments  
0            84.8          NaN       39.3        36.3           1        0  
1            88.9          NaN       37.9        34.9           1        0  
2            85.8          NaN       37.5        35.9           1        0  
3            82.9          NaN       39.8        35.6           1        0  
4            88.1          NaN       36.8        34.8           1        0

In [17]:
X = dataset.iloc[:, 8:9]


Great, we have fill the Comments column with zero values, Now, let's fill the FuelEconomy column with the mean value of the same column.

In [18]:
# Replacing values of NaN in the FuelEconomy column by the mean value of the same column

X = dataset.iloc[:, 0:13]

# Taking care of missing data
from sklearn.preprocessing import Imputer

imputer = Imputer(missing_values = 'NaN', strategy = 'mean', axis = 0)
imputer = imputer.fit(X[[8]])
X[[8]] = imputer.transform(X[[8]])



In [19]:
#renamimg our file
dataset = X

# saving our file locally
dataset.to_csv("Time-travel.csv", mode = 'w', index=False)

In [20]:
df = pd.read_csv('Time-travel.csv')
df.head()

Date StartTime  DayOfWeek  GoingTo  Distance  MaxSpeed  AvgSpeed  \
0  2012-01-06     16:37          4        1     51.29     127.4      78.3   
1  2012-01-06     08:20          4        0     51.63     130.3      81.8   
2  2012-01-04     16:17          2        1     51.27     127.4      82.0   
3  2012-01-04     07:53          2        0     49.17     132.3      74.2   
4  2012-01-03     18:57          1        1     51.15     136.2      83.4   

   AvgMovingSpeed  FuelEconomy  TotalTime  MovingTime  Take407All Comments  
0            84.8     8.690591       39.3        36.3           1        0  
1            88.9     8.690591       37.9        34.9           1        0  
2            85.8     8.690591       37.5        35.9           1        0  
3            82.9     8.690591       39.8        35.6           1        0  
4            88.1     8.690591       36.8        34.8           1        0

Great!  Now the data pre-processing it's over and we can start the modeling.